stan/math/prim/core/profiling.hpp

```diff
@@ -0,0 +1,67 @@
+#ifndef STAN_MATH_PRIM_CORE_PROFILING_HPP
+#define STAN_MATH_PRIM_CORE_PROFILING_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/err.hpp>
+#include <iostream>
+#include <sstream>
+#include <thread>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+template <typename T, require_all_not_var_t<T>* = nullptr>
+inline void profile_start(std::string name, profile_map& profiles) {
```

Comment:
I think passing profile names by `const char*` will be lower overhead. I know once the error messages got changed from `const char*` to `std::string` and it caused slowdowns cause the constructors or something.

Reply:
Agreed that const char* has a lower overhead. The question is more how much do we care about overhead in profiling? std::string is safer and has more C++ lib function to work with it.